File: coins.py
from flask import Blueprint, request, jsonify
import jwt
import uuid
from datetime import datetime
import os
from store import store
from routes.auth import token_required

coins = Blueprint('coins', __name__)

@coins.route('/packages', methods=['GET'])
def get_packages():
    try:
        packages = [
            {
                'id': 1,
                'name': '100 Z Coins',
                'coins': 100,
                'price': 999,
                'description': 'Basic package for starters'
            },
            {
                'id': 2,
                'name': '500 Z Coins',
                'coins': 500,
                'price': 4499,
                'description': 'Most popular package'
            },
            {
                'id': 3,
                'name': '1000 Z Coins',
                'coins': 1000,
                'price': 7999,
                'description': 'Best value package'
            }
        ]
        return jsonify(packages)
    except Exception as error:
        print('Get packages error:', str(error))
        return jsonify({'message': 'Error fetching packages'}), 500

@coins.route('/purchase', methods=['POST'])
@token_required
def purchase_coins(current_user):
    try:
        data = request.json
        package_id = data.get('packageId')
        
        if not package_id:
            return jsonify({'message': 'Package ID is required'}), 400

        # Find package
        packages = get_packages().json
        package = next((p for p in packages if p['id'] == package_id), None)
        
        if not package:
            return jsonify({'message': 'Package not found'}), 404

        # Update user coins
        current_user['coins'] += package['coins']
        store.users[current_user['userId']] = current_user

        # Create transaction record
        transaction_id = str(uuid.uuid4())
        transaction = {
            'id': transaction_id,
            'userId': current_user['userId'],
            'packageId': package_id,
            'coins': package['coins'],
            'amount': package['price'],
            'status': 'completed',
            'createdAt': datetime.now().isoformat()
        }

        # Store transaction
        if current_user['userId'] not in store.transactions:
            store.transactions[current_user['userId']] = {}
        store.transactions[current_user['userId']][transaction_id] = transaction

        return jsonify({
            'transactionId': transaction_id,
            'coins': current_user['coins'],
            'status': transaction['status']
        })

    except Exception as error:
        print('Purchase error:', str(error))
        return jsonify({'message': 'Error processing purchase'}), 500

@coins.route('/transaction/<transaction_id>', methods=['GET'])
@token_required
def get_transaction_status(current_user, transaction_id):
    try:
        user_transactions = store.transactions.get(current_user['userId'], {})
        transaction = user_transactions.get(transaction_id)
        
        if not transaction:
            return jsonify({'message': 'Transaction not found'}), 404

        return jsonify(transaction)

    except Exception as error:
        print('Get transaction status error:', str(error))
        return jsonify({'message': 'Error fetching transaction status'}), 500

@coins.route('/history', methods=['GET'])
@token_required
def get_transaction_history(current_user):
    try:
        user_transactions = store.transactions.get(current_user['userId'], {})
        transactions = list(user_transactions.values())
        transactions.sort(key=lambda x: x['createdAt'], reverse=True)
        return jsonify(transactions)

    except Exception as error:
        print('Get transaction history error:', str(error))
        return jsonify({'message': 'Error fetching transaction history'}), 500 

File: coins.js
const express = require('express');
const router = express.Router();
const jwt = require('jsonwebtoken');
const { v4: uuidv4 } = require('uuid');
const store = require('../store');

// Get coin packages
router.get('/packages', (req, res) => {
    const packages = [
        {
            id: '1',
            name: '1000 Z Coins',
            coins: 1000,
            price: 999,
            discount: 0
        },
        {
            id: '2',
            name: '2500 Z Coins',
            coins: 2500,
            price: 2249,
            discount: 10
        },
        {
            id: '3',
            name: '5000 Z Coins',
            coins: 5000,
            price: 4499,
            discount: 15
        }
    ];
    res.json(packages);
});

// Purchase coins
router.post('/purchase', async (req, res) => {
    try {
        const token = req.headers.authorization?.split(' ')[1];
        if (!token) {
            return res.status(401).json({ message: 'No token provided' });
        }

        const { packageId } = req.body;
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        const user = store.users.get(decoded.userId);

        if (!user) {
            return res.status(404).json({ message: 'User not found' });
        }

        // Find package
        const packages = await router.get('/packages').json;
        const package = packages.find(p => p.id === packageId);
        if (!package) {
            return res.status(404).json({ message: 'Package not found' });
        }

        // Update user coins
        user.coins += package.coins;
        store.users.set(decoded.userId, user);

        // Create transaction record
        const transaction = {
            id: uuidv4(),
            userId: user.userId,
            packageId,
            coins: package.coins,
            amount: package.price,
            status: 'completed',
            createdAt: new Date()
        };

        // Store transaction
        if (!store.transactions.has(user.userId)) {
            store.transactions.set(user.userId, new Map());
        }
        store.transactions.get(user.userId).set(transaction.id, transaction);

        res.json({
            transactionId: transaction.id,
            coins: user.coins,
            status: transaction.status
        });
    } catch (error) {
        console.error('Purchase error:', error);
        res.status(500).json({ message: 'Error processing purchase' });
    }
});

// Get transaction status
router.get('/transaction/:transactionId', async (req, res) => {
    try {
        const token = req.headers.authorization?.split(' ')[1];
        if (!token) {
            return res.status(401).json({ message: 'No token provided' });
        }

        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        const userTransactions = store.transactions.get(decoded.userId);
        const transaction = userTransactions?.get(req.params.transactionId);

        if (!transaction) {
            return res.status(404).json({ message: 'Transaction not found' });
        }

        res.json({
            id: transaction.id,
            status: transaction.status,
            coins: transaction.coins,
            amount: transaction.amount,
            createdAt: transaction.createdAt
        });
    } catch (error) {
        console.error('Transaction status error:', error);
        res.status(500).json({ message: 'Error fetching transaction status' });
    }
});

// Get transaction history
router.get('/history', async (req, res) => {
    try {
        const token = req.headers.authorization?.split(' ')[1];
        if (!token) {
            return res.status(401).json({ message: 'No token provided' });
        }

        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        const userTransactions = store.transactions.get(decoded.userId);
        
        if (!userTransactions) {
            return res.json([]);
        }

        const transactions = Array.from(userTransactions.values())
            .sort((a, b) => b.createdAt - a.createdAt);

        res.json(transactions);
    } catch (error) {
        console.error('Transaction history error:', error);
        res.status(500).json({ message: 'Error fetching transaction history' });
    }
});

module.exports = router; 